### ISSUE TITLE
Inconsistent Behaviour of Highlighter When Filling Is Activated

### ISSUE THREAD

**Issue by TheTrueBrot** [xournalpp/xournalpp#2904]
**Affects versions :**
 - Kubuntu 20.10 (5.8.0-44-generic)
 - Xournal++ 1.1.0+dev libgtk: 3.24.23
 - Installation method: PPA (I think)
 - I also remember seeing the problem on a windows system a few months ago, but currently don't have access to a windows system.

**Describe the bug**
The highlighter combined with the fill function sometimes produces borders and sometimes not, depending on the color selected.

**To Reproduce**
1. Activate the highlighter and filler
2. Draw a cirlce with the bright green -> You won't get a border.
3. Draw a circle with the dark green -> You will get a border.

**Expected behavior**
No boundary would be best, but if there is a reason for these boundaries, it would make sense for all highlighters to create a boundary. In short, the behaviour should be consistent.

**Screenshots of Problem**
Example with the rectangle tool:
![ezgif-2-ff0e13ed5ce1](https://user-images.githubusercontent.com/14249961/109974825-3fdc4300-7cfa-11eb-8cb5-31f603f43661.gif)

Example with hand-drawn circle:
![ezgif-2-bb3f2f646dfb](https://user-images.githubusercontent.com/14249961/109975100-8e89dd00-7cfa-11eb-8784-922944777610.gif)

**Additional context**
Maybe this is feature I don't understand :smile:. Feel free to close the issue if this is the case, but even then I think it would be great to have a option to make this consistent (in other words turn the borders off).

**Comment by rolandlo** [comment]
I agree, it looks strange with the border for the highlighter.

**Comment by Technius** [comment]
I think this may have to do with the combination of cairo operators used.

**Comment by mjg** [comment]
I think this is a cairo glitch handling the MULTIPLY operator. (All is fine with OVER in a patched xournalpp.) If you export to PDF you will find that different pdf viewers display it differently. cairo-based (evince) display the exact same glitch. mupdf-based (mupdf, zathura) show a very tiny white line instead of the border, but for all colours.

pdf2ps (gs) transforms this pdf into a ps which looks OK (pdf2ps renders tranparancy to embedded images).

Opening that pdf in scribus it looks perfect but I can reproduce the glitch by modifying the objects. In scribus I see for each highlighter filled rectangle:

- a group with transparancy 50% and mode MULTIPLY
- inside the group a polygon which has the same line (i.e. stroke) and fill colour at 100% opacity (0% transparent)
- the line has mode normal (which is OVER for cairo), the fill has mode MULTIPLY

If I change the line to mode MULTIPLY then I get the same glitches as reported here.

Funnily enough, an svg export sets the operator to MULTIPLY both on the fill and the stroke, and then again when using that surface, and yet inkscape (and others) displays it correctly.

I guess that MULTIPLY operator is not implemented right everywhere - possiblly the ordering or even an overflow in the formula. The white hair line maybe a rounding issue?

Stomped.

**Comment by mjg** [comment]
Maybe an interesting note: draw, say, a rectangle with the higlighter, then select it with the select tool and switch fill on and off. You see that the "extra border" is exactly where the fill rectangle intersects with the stroke: the stroke is "left and right" (half a strokewidth each) of the (mathematically zero width) path, the fill inside, resulting in an intersection of half a strokewidth.

**Comment by bhennion** [comment]
This is really what is to be expected for a `CAIRO_OPERATOR_MULTIPLY` (see [here](https://www.cairographics.org/operators/)). This operator multiplies the `RGBA` values (normalized between 0 and 1) of the background with that of the foreground. In our case, the filling is painted before the stroke itself, so the resulting color in the intersection of the stroke and its filling may differ from the rest.

Two examples:

- Yellow: RGB = (1.0, 1.0, 0.0) so Yellow over Yellow gives (1.0 * 1.0, 1.0 * 1.0, 0.0 * 0.0) = (1.0, 1.0, 0.0) = Yellow again. We don't see a border.
- Orange: RGB = (1.0, 0.5, 0.0) so Orange over Orange gives (1.0 * 1.0, 0.5 * 0.5, 0.0 * 0.0) = (1.0, 0.25, 0.0) = some darker orange. We see a border.

(The transparency is applied later, so not multiplied).

**Comment by mjg** [comment]
I agree that this is expected from the implementation using `MULTIPLY`, on the other hand the exported PDF displays differently in different viewers, and fills for the stroke tool behave as (user) expected.

More importantly: This does not seem to be expected highlighter behaviour. If you think of applying semi-transparent colour to paper then things should get less transparent with every layer that you apply. On the other hand, for the typical digital highlighter you might not even want a thicker border. Both behaviours would be fine, but not a mix of these behaviours for different colors.

In summary, `MULTIPLY` might not be the best choice, but the analysis in my previous comment indicates that it's really the interaction of different modes on strokes and fills and the enclosing group which interact, potentially differently in different viewers.

**Comment by bhennion** [comment]
Yes, of course it should be corrected! I actually wrote a patch for that yesterday. I'll clean it up and make a PR ASAP.
We don't need to drop `MULTIPLY`, just adapt the use of `cairo_push_group` a little bit.

**Comment by bhennion** [comment]
See #3355 for a PR fixing this
